text/0779-first-class-component-templates.md

```diff
@@ -0,0 +1,1401 @@
+
+
+#### Class body
+
+The compilation output with a class-backed component is similar, but instead of using `templateOnlyComponent`, it uses the backing class. Given this component:
+
+```js
+import Component from '@glimmer/component';
+import { tracked } from '@glimmer/tracking';
+
+class SetUsername extends Component {
+  @tracked name = '';
+
+  updateName = ({ target: { value } }) => {
+    this.name = value;
+  }
+
+  saveName = (submitEvent) => {
+    submitEvent.preventDefault();
+    this.args.onSaveName(this.name);
+  };
+
+  <template>
+    <form {{on "submit" this.saveName}}>
+      <label for='name'>Set username:</label>
+      <input
+        id='name'
+        value={{this.value}}
+        {{on "input" this.updateName}}
+      />
+      <button type='submit' disabled={{eq this.value.length 0}}>
+        Generate
+      </button>
+    </form>
+  </template>
+}
+```
+
+The compiled output is:
+
+```js
+import Component from '@glimmer/component';
+import { tracked } from '@glimmer/tracking';
+import { precompileTemplate } from '@ember/template-compilation';
+import { setComponentTemplate } from '@ember/component';
+
+class SetUsername extends Component {
+  @tracked name = '';
+
+  updateName = ({ target: { value } }) => {
+    this.name = value;
+  }
+
+  saveName = (submitEvent) => {
+    submitEvent.preventDefault();
+    this.args.onSaveName(this.name);
+  };
+}
+
+setComponentTemplate(
+  precompileTemplate`
+    <form {{on "submit" this.saveName}}>
+      <label for='name'>Set username:</label>
+      <input
+        id='name'
+        value={{this.value}}
+        {{on "input" this.updateName}}
+      />
+      <button type='submit' disabled={{eq this.value.length 0}}>
+        Generate
+      </button>
+    </form>
+  `
+  ),
+  SetUsername
+)
+```
+
+##### Private class fields
+
+In the present design of the template compilation primitives, a template cannot access private fields from the backing class. That is, the following *will not work*:
+
+```js
+import Component from '@glimmer/component';
+
+class Example extends Component {
+  #aField = true;
+
+  <template>
+    <p>The value of the field is: {{this.#aField}}</p>
+  </template>
+}
+```
+
+That is because the compilation output does *not* embed the template in the class' body in any way, but instead associates it *externally* to the class—but private class fields are only accessible within the body of the class itself, per the ECMAScript spec. While we could invest time to change the implementation to avoid this, it is not generally a problem. Because we do not interact with component class instances directly—only through the template layer—we do not actually *need* private class fields.
+
+Even, this is a real gap, which we could address in a future RFC. Notably, it is *not specific to this proposal*, but applies to *all* proposals built on the current primitives.
```

Comment:
I find the Even here confusing, you might mean this?

```suggestion
This is a real gap, which we could address in a future RFC. Notably, it is *not specific to this proposal*, but applies to *all* proposals built on the current primitives.
```

Reply:
Yeah, I think I was going for "Even so" or something originally and then changed sentence structure. Much obliged!